framework/include/userobject/PropertyReadFile.h

```diff
@@ -0,0 +1,142 @@
+  virtual ~PropertyReadFile() {}
+
+  virtual void initialize() {}
+  virtual void execute() {}
+  virtual void finalize() {}
+
+  /**
+   * This function reads the data from file
+   */
+  void readData();
+
+  /**
+   * This function generates voronoi tesselation center points
+   * Presently random generated
+   */
+  virtual void initVoronoiCenterPoints();
+
+  /**
+   * This function assign property data to elements
+   * @param elem the element to get data for
+   * @param prop_num the column index of the property we want to retrieve
+   */
+  Real getData(const Elem * elem, unsigned int prop_num) const;
+
+  /**
+   * This function assign properties to element read from file with element based properties
+   * @param elem the element to get data for
+   * @param prop_num the column index of the property we want to retrieve
+   */
+  Real getElementData(const Elem * elem, unsigned int prop_num) const;
+
+  /**
+   * This function assign properties to read from file with node based properties
+   * @param node the node to get the data for
+   * @param prop_num the column index of the property we want to retrieve
+   */
+  Real getNodeData(const Node * node, unsigned int prop_num) const;
+
+  /**
+   * This function assign properties to element read from file with nearest neighbor / grain based
+   * properties Voronoi centers distribution in the RVE can be Periodic or non-periodic (default)
+   * @param centroid the centroid of the element to get data for
+   * @param prop_num the column index of the property we want to retrieve
+   */
+  Real getVoronoiData(const Point centroid, unsigned int prop_num) const;
+
+  /**
+   * This function assigns properties to elements read from file with block  based properties
+   * @param elem the element to get data for
+   * @param prop_num the column index of the property we want to retrieve
+   */
+  Real getBlockData(const Elem * elem, unsigned int prop_num) const;
+
+  /**
+   * This function calculates minimum distance between 2 points
+   * considering periodicity of the simulation volume
+   */
+  Real minPeriodicDistance(Point, Point) const;
+
+  /**
+   * Returns the ordering of data expected in the CSV file
+   */
+  ReadTypeEnum getReadType() const { return _read_type; }
+
+protected:
+  /// Name of file containing property values
+  const std::string _prop_file_name;
+  /// Use DelimitedFileReader to read and store data from file
+  MooseUtils::DelimitedFileReader _reader;
+  /// Number of properties in a row
+  const unsigned int _nprop;
+  /// Number of grains (for property read based on grains)
+  const unsigned int _nvoronoi;
+  /// Number of blocks (for property read based on blocks)
+  const unsigned int _nblock;
+  /// Type of read - element, grain, or block (TODO: Make const once apps are patched)
+  ReadTypeEnum _read_type;
+  /// Whether to use a random tesselation for the Voronoi/grain type
+  const bool _use_random_tesselation;
```

Comment:
? 

This class should focus on reading data from a file. Why is there functionality to perform some random tessellation?!

Reply:
because that s what it used to do in Tensor Mecha

It s doing a random tessellation and reading data, assigning it to that random tessellation. It's a general functionality. Probably a rare one outside of TM

Reply:
The "periodic grain option" could also be shed. But it doesnt hurt anything, and "periodicity" or "randomness" are not concepts specific to materials stuff